File: modules/networking-api/src/main/java/com/roxiemobile/networkingapi/network/http/PersistentCookieStore.java
package com.roxiemobile.networkingapi.network.http;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.androidcommons.logging.Logger;
import com.roxiemobile.androidcommons.util.CollectionUtils;
import com.roxiemobile.androidcommons.util.IOUtils;
import com.roxiemobile.networkingapi.util.CookieUtils;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public final class PersistentCookieStore implements CookieStore
{
// MARK: - Construction

    public PersistentCookieStore(Context context, HttpCookie[] cookies) {
        this(context, (cookies != null) ? Arrays.asList(cookies) : null);
    }

    public PersistentCookieStore(Context context, List<HttpCookie> cookies) {
        this(context);

        // Add cookies to CookieStore
        if (CollectionUtils.isNotEmpty(cookies)) {
            for (HttpCookie cookie : cookies) {
                add(CookieUtils.cookiesUri(cookie), cookie);
            }
        }
    }

    public PersistentCookieStore(Context context, CookieStore otherCookieStore) {
        this(context);

        // Copy cookies from other CookieStore
        if (otherCookieStore != null) {
            for (URI uri : otherCookieStore.getURIs()) {
                for (HttpCookie cookie : otherCookieStore.get(uri)) {
                    add(uri, cookie);
                }
            }
        }
    }

    public PersistentCookieStore(Context context) {
        // Init instance variables
        mSharedPreferences = context.getSharedPreferences(SP_COOKIE_STORE, Context.MODE_PRIVATE);
        loadAllFromPersistence();
    }

// MARK: - Methods

    public synchronized void add(URI uri, @NonNull HttpCookie cookie) {
        Guard.notNull(cookie, "cookie is null");

        uri = CookieUtils.cookiesUri(uri);
        List<HttpCookie> cookies = mMap.get(uri);
        if (cookies == null) {
            cookies = new ArrayList<>();
            mMap.put(uri, cookies);
        }
        else {
            cookies.remove(cookie);
        }
        cookies.add(cookie);
        saveToPersistence(uri, cookie);
    }

    public synchronized List<HttpCookie> get(@NonNull URI uri) {
        Guard.notNull(uri, "uri is null");

        List<HttpCookie> result = new ArrayList<>();
        List<HttpCookie> cookiesToRemoveFromPersistence = new ArrayList<>();
        final Date date = new Date();

        // Get cookies associated with given URI. If none, returns an empty list
        List<HttpCookie> cookiesForUri = mMap.get(uri);
        if (cookiesForUri != null) {
            for (Iterator<HttpCookie> i = cookiesForUri.iterator(); i.hasNext(); ) {
                HttpCookie cookie = i.next();
                if (cookie.hasExpired(date)) {
                    cookiesToRemoveFromPersistence.add(cookie);
                    i.remove(); // remove expired cookies
                }
                else {
                    result.add(cookie);
                }
            }
        }

        // Get all cookies that domain matches the URI
        for (Map.Entry<URI, List<HttpCookie>> entry : mMap.entrySet()) {
            if (uri.equals(entry.getKey())) {
                continue; // skip the given URI; we've already handled it
            }

            List<HttpCookie> entryCookies = entry.getValue();
            for (Iterator<HttpCookie> i = entryCookies.iterator(); i.hasNext(); ) {
                HttpCookie cookie = i.next();
                if (!HttpCookie.domainMatches(cookie.getDomain(), uri.getHost())) {
                    continue;
                }
                if (cookie.hasExpired(date)) {
                    cookiesToRemoveFromPersistence.add(cookie);
                    i.remove(); // remove expired cookies
                }
                else if (!result.contains(cookie)) {
                    result.add(cookie);
                }
            }
        }

        if (!cookiesToRemoveFromPersistence.isEmpty()) {
            removeFromPersistence(uri, cookiesToRemoveFromPersistence);
        }

        return Collections.unmodifiableList(result);
    }

    public synchronized List<HttpCookie> getCookies() {
        List<HttpCookie> result = new ArrayList<>();
        List<HttpCookie> cookiesToRemoveFromPersistence = new ArrayList<>();
        final Date date = new Date();

        for (URI uri : mMap.keySet()) {
            for (Iterator<HttpCookie> i = mMap.get(uri).iterator(); i.hasNext(); ) {
                HttpCookie cookie = i.next();
                if (cookie.hasExpired(date)) {
                    cookiesToRemoveFromPersistence.add(cookie);
                    i.remove(); // remove expired cookies
                }
                else if (!result.contains(cookie)) {
                    result.add(cookie);
                }
            }

            if (!cookiesToRemoveFromPersistence.isEmpty()) {
                removeFromPersistence(uri, cookiesToRemoveFromPersistence);
            }
        }
        return Collections.unmodifiableList(result);
    }

    public synchronized List<URI> getURIs() {
        List<URI> result = new ArrayList<>(mMap.keySet());
        result.remove(null); // sigh
        return Collections.unmodifiableList(result);
    }

    public synchronized boolean remove(URI uri, @NonNull HttpCookie cookie) {
        Guard.notNull(cookie, "cookie is null");

        List<HttpCookie> cookies = mMap.get(CookieUtils.cookiesUri(uri));
        boolean result = false;

        if (cookies != null) {
            result = cookies.remove(cookie);
            if (cookies.isEmpty()) {
                mMap.remove(uri);
            }
            if (result) {
                removeFromPersistence(uri, cookie);
            }
        }

        return result;
    }

    public synchronized boolean removeAll() {
        boolean result = !mMap.isEmpty();
        mMap.clear();
        removeAllFromPersistence();
        return result;
    }

// MARK: - Private Methods

    private void loadAllFromPersistence() {
        mMap = new HashMap<>();

        Map<String, ?> allPairs = mSharedPreferences.getAll();
        for (Entry<String, ?> entry : allPairs.entrySet()) {
            String[] uriAndName = entry.getKey().split(SP_KEY_DELIMITER_REGEX, 2);
            try {
                URI uri = new URI(uriAndName[0]);
                String encodedCookie = (String) entry.getValue();
                HttpCookie cookie = (HttpCookie) IOUtils.decodeObject(encodedCookie);

                List<HttpCookie> targetCookies = mMap.get(uri);
                if (targetCookies == null) {
                    targetCookies = new ArrayList<>();
                    mMap.put(uri, targetCookies);
                }
                // Repeated cookies cannot exist in persistence
                // targetCookies.remove(cookie)
                targetCookies.add(cookie);
            }
            catch (URISyntaxException ex) {
                Logger.w(TAG, ex);
            }
        }
    }

    private void saveToPersistence(URI uri, HttpCookie cookie) {
        SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putString(uri.toString() + SP_KEY_DELIMITER + cookie.getName(), IOUtils.encodeObject(cookie));
        editor.apply();
    }

    private void removeFromPersistence(URI uri, List<HttpCookie> cookiesToRemove) {
        SharedPreferences.Editor editor = mSharedPreferences.edit();
        for (HttpCookie cookieToRemove : cookiesToRemove) {
            editor.remove(uri.toString() + SP_KEY_DELIMITER + cookieToRemove.getName());
        }
        editor.apply();
    }

    private void removeFromPersistence(URI uri, HttpCookie cookieToRemove) {
        SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.remove(uri.toString() + SP_KEY_DELIMITER + cookieToRemove.getName());
        editor.apply();
    }

    private void removeAllFromPersistence() {
        mSharedPreferences.edit().clear().apply();
    }

// MARK: - Constants

    private static final String TAG = PersistentCookieStore.class.getSimpleName();

    // Persistence
    private static final String SP_COOKIE_STORE = "cookieStore";
    private static final String SP_KEY_DELIMITER = "|"; // Unusual char in URL
    private static final String SP_KEY_DELIMITER_REGEX = "\\" + SP_KEY_DELIMITER;

// MARK: - Variables

    private SharedPreferences mSharedPreferences;

    // This map may have null keys!
    private Map<URI, List<HttpCookie>> mMap = new HashMap<>();

}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/http/PersistentCookieStore.kt
@file:Suppress("UsePropertyAccessSyntax", "unused")

package com.roxiemobile.networkingapi.network.http

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import com.roxiemobile.androidcommons.logging.Logger
import com.roxiemobile.androidcommons.util.IOUtils
import com.roxiemobile.kotlincommons.util.Base62
import com.roxiemobile.networkingapi.network.http.util.CookieUtils
import com.roxiemobile.networkingapi.network.http.util.LinkedMultiValueMap
import com.roxiemobile.networkingapi.network.http.util.MultiValueMap
import kotlinx.collections.immutable.toImmutableList
import java.io.StreamCorruptedException
import java.net.URI
import java.util.Date

class PersistentCookieStore: CookieStore {

// MARK: - Construction

    constructor(context: Context) {
        _sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE)
        loadFromSharedPreferences()
    }

    constructor(context: Context, cookies: List<HttpCookie>?): this(context) {
        if (cookies != null) {
            for (cookie in cookies) {
                add(CookieUtils.cookieLink(cookie), cookie)
            }
        }
    }

    constructor(context: Context, otherCookieStore: CookieStore?): this(context) {
        if (otherCookieStore != null) {
            for (link in otherCookieStore.getURIs()) {
                for (cookie in otherCookieStore[link]) {
                    add(link, cookie)
                }
            }
        }
    }

// MARK: - Methods

    @Synchronized
    override fun add(uri: URI?, cookie: HttpCookie) {
        val link = CookieUtils.cookieLink(uri)

        val cookies: MutableList<HttpCookie>? = _cookies[link]
        cookies?.remove(cookie)

        saveToSharedPreferences(link, cookie)
        _cookies.add(link, cookie)
    }

    @Synchronized
    override fun get(uri: URI?): List<HttpCookie> {
        val link = CookieUtils.cookieLink(uri)

        val outCookies = mutableListOf<HttpCookie>()
        val expiredCookies = mutableListOf<HttpCookie>()
        val date = Date()

        // Get cookies associated with given link. If none, returns an empty list
        _cookies[link]?.let { cookies ->
            val iterator = cookies.iterator()

            while (iterator.hasNext()) {
                val cookie = iterator.next()

                if (cookie.hasExpired(date)) {
                    iterator.remove() // remove expired cookies
                    expiredCookies.add(cookie)
                }
                else {
                    outCookies.add(cookie)
                }
            }
        }

        // Get all cookies that domain matches the link
        if (link != null) {
            for (entry in _cookies.entries) {

                if (link == entry.key) {
                    continue // skip the given link; we've already handled it
                }

                val entryCookies: MutableList<HttpCookie> = entry.value
                val iterator = entryCookies.iterator()

                while (iterator.hasNext()) {
                    val cookie = iterator.next()

                    if (HttpCookie.domainMatches(cookie.domain, link.host)) {

                        if (cookie.hasExpired(date)) {
                            iterator.remove() // remove expired cookies
                            expiredCookies.add(cookie)
                        }
                        else if (!outCookies.contains(cookie)) {
                            outCookies.add(cookie)
                        }
                    }
                }
            }
        }

        if (expiredCookies.isNotEmpty()) {
            removeFromSharedPreferences(link, expiredCookies)
            expiredCookies.clear()
        }

        return outCookies.toImmutableList()
    }

    @Synchronized
    override fun getCookies(): List<HttpCookie> {

        val outCookies = mutableListOf<HttpCookie>()
        val expiredCookies = mutableListOf<HttpCookie>()
        val date = Date()

        for (link in _cookies.keys) {

            _cookies[link]?.let { cookies ->
                val iterator = cookies.iterator()

                while (iterator.hasNext()) {
                    val cookie = iterator.next()

                    if (cookie.hasExpired(date)) {
                        iterator.remove() // remove expired cookies
                        expiredCookies.add(cookie)
                    }
                    else if (!outCookies.contains(cookie)) {
                        outCookies.add(cookie)
                    }
                }
            }

            if (expiredCookies.isNotEmpty()) {
                removeFromSharedPreferences(link, expiredCookies)
                expiredCookies.clear()
            }
        }

        return outCookies.toImmutableList()
    }

    @Synchronized
    override fun getURIs(): List<URI> {
        val links = _cookies.keys.filterNotNull()
        return links.toImmutableList()
    }

    @Synchronized
    override fun remove(uri: URI?, cookie: HttpCookie): Boolean {

        val link = CookieUtils.cookieLink(uri)
        var result = false

        _cookies[link]?.let { cookies ->
            result = cookies.remove(cookie)

            if (cookies.isEmpty()) {
                // Remove key without values
                _cookies.remove(link)
            }

            if (result) {
                removeFromSharedPreferences(link, cookie)
            }
        }

        return result
    }

    @Synchronized
    override fun removeAll(): Boolean {
        return _cookies.isNotEmpty().also {
            clearSharedPreferences()
            _cookies.clear()
        }
    }

// MARK: - Private Methods

    private fun loadFromSharedPreferences() {
        _cookies.clear()

        for (entry in _sharedPreferences.all) {
            try {

                val storageKey = decodeKey(entry.key)
                val decodedObject = IOUtils.decodeObject(entry.value as String)

                (decodedObject as? HttpCookie)?.let { cookie ->
                    _cookies.add(storageKey.link, cookie)
                }
            }
            catch (ex: StreamCorruptedException) {
                Logger.w(TAG, ex)
            }
        }
    }

    private fun saveToSharedPreferences(link: URI?, cookie: HttpCookie) {
        _sharedPreferences.edit {
            val storageKey = encodeKey(StorageKey(link, cookie.name))
            val value = IOUtils.encodeObject(cookie)
            putString(storageKey, value)
        }
    }

    private fun removeFromSharedPreferences(link: URI?, cookies: List<HttpCookie>) {
        _sharedPreferences.edit {
            cookies.forEach { cookie ->
                val storageKey = encodeKey(StorageKey(link, cookie.name))
                remove(storageKey)
            }
        }
    }

    private fun removeFromSharedPreferences(link: URI?, cookie: HttpCookie) {
        _sharedPreferences.edit {
            val storageKey = encodeKey(StorageKey(link, cookie.name))
            remove(storageKey)
        }
    }

    private fun clearSharedPreferences() {
        _sharedPreferences.edit {
            clear()
        }
    }

    private fun encodeKey(storageKey: StorageKey): String {

        val components = listOf(
            storageKey.link.toString(),
            storageKey.cookieName,
        )

        return components.joinToString(KEY_DELIMITER) {
            Base62.SHARED.encode(it.toByteArray()).toString()
        }
    }

    @Throws(StreamCorruptedException::class)
    private fun decodeKey(key: String): StorageKey {
        try {
            val components = key.split(KEY_DELIMITER, limit = 2).map {
                Base62.SHARED.decode(it.toByteArray()).toString()
            }

            val linkText = components[0]
            val link = URI.create(if (linkText != "null") linkText else null)

            return StorageKey(link, components[1])
        }
        catch (ex: Exception) {
            throw StreamCorruptedException()
        }
    }

// MARK: - Inner Types

    private data class StorageKey(val link: URI?, val cookieName: String)

// MARK: - Companion

    companion object {

        private val TAG = PersistentCookieStore::class.java.simpleName

        private const val SHARED_PREFS_NAME = "com.roxiemobile.networkingapi.network.http.cookieStore"
        private const val KEY_DELIMITER = ":"
    }

// MARK: - Variables

    private val _sharedPreferences: SharedPreferences

    // NOTE: This map may have null keys!
    private val _cookies: MultiValueMap<URI?, HttpCookie> = LinkedMultiValueMap()
}
